dask/task_spec.py

```diff
@@ -0,0 +1,905 @@
+        if k is None:
+            k = ("anom", next(_anom_count))
+
+        func, args = arg[0], arg[1:]
+        if isinstance(func, SubgraphCallable):
+            # FIXME: This parsing is inlining tasks of the subgraph, i.e. it can
+            # multiply work if the subgraph has keys that are reused in the
+            # subgraph
+            subgraph = func
+            all_keys_inner = _MultiSet(all_keys, set(subgraph.inkeys), subgraph.dsk)
+            if subgraph.name in cache:
+                func = cache[subgraph.name]
+            else:
+                sub_dsk = convert_old_style_dsk(
+                    subgraph.dsk, all_keys_inner, only_refs, cache
+                )
+                cache[subgraph.name] = func = sub_dsk[subgraph.outkey].inline(sub_dsk)
+                del sub_dsk
+
+            new_args_l = map(partial(inner, all_keys=all_keys_inner), args)
+
+            func2 = func.inline(dict(zip(subgraph.inkeys, new_args_l)))
+            dct = {k: Alias(k) for k in func2.dependencies}
+            return Task(key=k, func=func2, args=[dct])
+        else:
+            new_args = map(inner_args, args)
+            if only_refs:
+                # this can't be a literal since the literal wouldn't execute
+                # new_args in case there were runnable tasks or other literals
+                # See test_convert_task_only_ref
+                return Task(k, identity, (func, *new_args))
+            return Task(key=k, func=func, args=tuple(new_args))
+    try:
+        if isinstance(arg, (bytes, int, float, str, tuple)) and arg in all_keys:
+            if arg in cache:
+                return cache[arg]
+            cache[arg] = rv = Alias(arg)
+            return rv
+    except TypeError:
+        # Unhashable
+        pass
+
+    if isinstance(arg, (list, tuple, set, frozenset)):
+        if is_namedtuple_instance(arg):
+            rv = _wrap_namedtuple_task(k, arg, inner_args)
+        else:
+            rv = SequenceOfTasks(k, map(inner_args, arg), typ=type(arg))
+        return rv.propagate_literal()
+    elif isinstance(arg, dict):
+        new_dsk = convert_old_style_dsk(
+            arg, all_keys=all_keys, only_refs=True, cache=cache
+        )
+
+        return DictOfTasks(None, new_dsk).propagate_literal()
+
+    elif isinstance(arg, WrappedKey):
+        return Alias(arg.key)
+    else:
+        return arg
+
+
+def convert_old_style_dsk(
+    dsk: Mapping,
+    all_keys: Container | None = None,
+    only_refs: bool = False,
+    cache: MutableMapping | None = None,
+):
+    if not all_keys:
+        all_keys = set(dsk)
+    new_dsk = {}
+    if cache is None:
+        cache = {}
+    for k, arg in dsk.items():
+        t = convert_old_style_task(k, arg, all_keys, only_refs, cache=cache)
+        if not only_refs and isinstance(t, Alias) and t.key == k:
+            continue
+        new_dsk[k] = t
+    new_dsk2 = {
+        k: v if isinstance(v, BaseTask) else LiteralTask(k, v)
```

Comment:
We can't avoid this second loop?

Reply:
perf wise that doesn't really matter (the only additional work we are doing here is initialize the for loop, instantiating the iterator, etc.)
The cost for the loop body is the same, i.e. we'll have to do the isinstance check for every task regardless of whether this is another loop or not.

The reason why this is here in the first place is because this ensures that after the conversion layer, every value in `dsk` is of type `BaseTask` which is not ensured by `convert_old_style_task`. I chose to not do this conversion in `convert_old_style_task` right away since `convert_old_style_task` is used recursively which would force us to wrap *every* object we encounter (which also works, this was my initial version but this is obviously slower)

Reply:
Thanks!